src/platforms/android/androidiaphandler.cpp

```diff
@@ -0,0 +1,344 @@
+  Q_UNUSED(thiz);
+
+  QJsonObject obj = AndroidUtils::getQJsonObjectFromJString(env, data);
+  if (!obj.contains("products")) {
+    logger.error() << "onSkuDetailsReceived - products entry expected.";
+    return;
+  }
+  QJsonArray products = obj["products"].toArray();
+  if (products.isEmpty()) {
+    logger.error() << "onSkuDetailsRecieved - no products found.";
+    return;
+  }
+  IAPHandler* iap = IAPHandler::instance();
+  static_cast<AndroidIAPHandler*>(iap)->updateProductsInfo(products);
+  iap->productsRegistrationCompleted();
+}
+
+// Call backs from JNI - Failures
+
+// static
+void AndroidIAPHandler::onBillingNotAvailable(JNIEnv* env, jobject thiz,
+                                              jstring data) {
+  Q_UNUSED(thiz);
+  QJsonObject billingResponse =
+      AndroidUtils::getQJsonObjectFromJString(env, data);
+  logger.info()
+      << "onBillingNotAvailable event occured"
+      << QJsonDocument(billingResponse).toJson(QJsonDocument::Compact);
+  IAPHandler* iap = IAPHandler::instance();
+  if (billingResponse["code"].toInt() == -99) {
+    // The billing service was disconnected.
+    // Lets try a reset if we need a subscription.
+    // TODO - This is speculative. I put it here because I got an inappropriate
+    // launch of the "Sign in to Play Store" window. But I'm not exactly sure
+    // how to trigger a billing service disconnected and I'm not sure what the
+    // right action should be.
+    MozillaVPN* vpn = MozillaVPN::instance();
+    if (vpn->user()->subscriptionNeeded()) {
+      vpn->reset(true);
+      return;
+    }
+  }
+  iap->stopSubscription();
+  iap->cancelProductsRegistration();
+  emit iap->billingNotAvailable();
+}
+
+// static
+void AndroidIAPHandler::onPurchaseAcknowledgeFailed(JNIEnv* env, jobject thiz,
+                                                    jstring data) {
+  Q_UNUSED(thiz);
+  QJsonObject json = AndroidUtils::getQJsonObjectFromJString(env, data);
+  logger.error() << "onPurchaseAcknowledgeFailed"
+                 << QJsonDocument(json).toJson(QJsonDocument::Compact);
+  IAPHandler* iap = IAPHandler::instance();
+  iap->stopSubscription();
+  emit iap->subscriptionNotValidated();
+}
+
+// static
+void AndroidIAPHandler::onSkuDetailsFailed(JNIEnv* env, jobject thiz,
+                                           jstring data) {
+  Q_UNUSED(thiz);
+  QJsonObject json = AndroidUtils::getQJsonObjectFromJString(env, data);
+  logger.error() << "onSkuDetailsFailed"
+                 << QJsonDocument(json).toJson(QJsonDocument::Compact);
+  IAPHandler* iap = IAPHandler::instance();
+  iap->stopSubscription();
+  emit iap->subscriptionFailed();
+}
+
+// static
+void AndroidIAPHandler::onSubscriptionFailed(JNIEnv* env, jobject thiz,
+                                             jstring data) {
+  Q_UNUSED(thiz);
+  QJsonObject json = AndroidUtils::getQJsonObjectFromJString(env, data);
+  logger.error() << "onSubscriptionFailed"
```

Comment:
hopefully, there are no sensitive data here and in the 3 previous methods.

Reply:
No, these are all returning error codes from Android IAP. It is the Serializable `BillingResultData` that comes back.